pkg/docker/image_id.go

```diff
@@ -0,0 +1,104 @@
+package docker
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/docker/distribution/reference"
+)
+
+const (
+	// Durations in seconds
+	oneHour  = 3600
+	oneDay   = oneHour * 24
+	oneMonth = oneDay * 28
+)
+
+type EmptyTag struct{}
+type NameTag string
+type DigestTag string
+
+type ImageTag interface {
+	CacheDuration() int64
+	Separator() string
+	String() string
+}
+
+func (EmptyTag) Separator() string    { return "" }
+func (EmptyTag) String() string       { return "" }
+func (EmptyTag) CacheDuration() int64 { return oneHour }
+
+func (NameTag) Separator() string { return ":" }
+func (n NameTag) String() string  { return string(n) }
+func (n NameTag) CacheDuration() int64 {
+	if n.String() == "latest" {
+		return oneHour
+	}
+	return oneDay
+}
+
+func (DigestTag) Separator() string    { return "@" }
+func (n DigestTag) String() string     { return string(n) }
+func (DigestTag) CacheDuration() int64 { return oneMonth }
+
+func ParseImageTag(s string) (string, ImageTag) {
+	// Check for a digest, and if not look for a tag
+	if strings.Contains(s, "@") {
+		tag_split := strings.Split(s, "@")
+		return tag_split[0], DigestTag(tag_split[1])
+	} else if strings.Contains(s, ":") {
+		tag_split := strings.Split(s, ":")
+		return tag_split[0], NameTag(tag_split[1])
+	}
+
+	return s, EmptyTag{}
+}
+
+type ImageID struct {
+	// e.g. ghcr.io
+	repository string
+
+	// e.g. ubuntu or organisation/ubuntu or organisation/user/ubuntu
+	name string
+
+	// e.g. latest or kinetic or @sha256:digest-string
+	tag ImageTag
+}
+
+func NewImageID(imageID string) (*ImageID, error) {
+	id := &ImageID{}
+
+	repo, err := reference.Parse(imageID)
+	if err != nil {
+		return nil, err
+	}
+
+	if named, ok := repo.(reference.Named); ok {
+		id.repository = reference.Domain(named)
+		id.name = reference.Path(named)
+	}
+
+	if digested, ok := repo.(reference.Digested); ok {
+		digest := fmt.Sprintf("sha256:%s", digested.Digest().Encoded())
```

Comment:
It's not always guaranteed to be `sha256` right? Can we call `Digest().Algorithm()` for this?

Reply:
Yup, you're right, it does support others but 256 is the canonical one, have changed.